File: sdk/go/client_test.go
package hnsx_test

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/hnsx-io/hnsx/sdk/go"
)

func newTestServer(t *testing.T) *httptest.Server {
	t.Helper()
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/api/v1/domains":
			if r.Method == http.MethodGet {
				w.Header().Set("Content-Type", "application/json")
				_ = json.NewEncoder(w).Encode(map[string]any{
					"items": []map[string]any{{"id": "customer-service", "version": "1.0.0", "status": "active"}},
					"total": 1,
				})
				return
			}
		case "/api/v1/sessions":
			if r.Method == http.MethodPost {
				var body map[string]any
				_ = json.NewDecoder(r.Body).Decode(&body)
				w.Header().Set("Content-Type", "application/json")
				_ = json.NewEncoder(w).Encode(map[string]any{
					"id":        "sess-123",
					"domain_id": body["domain_id"],
					"state":     "running",
					"trigger":   body["trigger"],
				})
				return
			}
		case "/api/v1/approvals/approve-1/approve":
			if r.Method == http.MethodPost {
				w.Header().Set("Content-Type", "application/json")
				_ = json.NewEncoder(w).Encode(map[string]any{"id": "approve-1", "status": "approved"})
				return
			}
		case "/api/v1/domains/missing":
			if r.Method == http.MethodGet {
				w.WriteHeader(http.StatusNotFound)
				_ = json.NewEncoder(w).Encode(map[string]any{"code": "DOMAIN_NOT_FOUND", "message": "not found"})
				return
			}
		}
		w.WriteHeader(http.StatusNotFound)
	}))
}

func TestClient_DomainsList(t *testing.T) {
	server := newTestServer(t)
	defer server.Close()

	client := hnsx.NewClient(server.URL)
	resp, err := client.Domains.List(context.Background(), 50, 0)
	if err != nil {
		t.Fatalf("list domains: %v", err)
	}
	if resp.Total != 1 || len(resp.Items) != 1 {
		t.Fatalf("unexpected response: %+v", resp)
	}
	if resp.Items[0].ID != "customer-service" {
		t.Fatalf("unexpected id: %s", resp.Items[0].ID)
	}
}

func TestClient_SessionsTrigger(t *testing.T) {
	server := newTestServer(t)
	defer server.Close()

	client := hnsx.NewClient(server.URL)
	session, err := client.Sessions.Trigger(context.Background(), "customer-service", map[string]any{"question": "hi"})
	if err != nil {
		t.Fatalf("trigger session: %v", err)
	}
	if session.ID != "sess-123" {
		t.Fatalf("unexpected session id: %s", session.ID)
	}
	if session.DomainID != "customer-service" {
		t.Fatalf("unexpected domain id: %s", session.DomainID)
	}
}

func TestClient_ApprovalsApprove(t *testing.T) {
	server := newTestServer(t)
	defer server.Close()

	client := hnsx.NewClient(server.URL)
	approval, err := client.Approvals.Approve(context.Background(), "approve-1")
	if err != nil {
		t.Fatalf("approve: %v", err)
	}
	if approval.Status != "approved" {
		t.Fatalf("unexpected status: %s", approval.Status)
	}
}

func TestClient_APIError(t *testing.T) {
	server := newTestServer(t)
	defer server.Close()

	client := hnsx.NewClient(server.URL)
	_, err := client.Domains.Get(context.Background(), "missing")
	apiErr, ok := err.(*hnsx.APIError)
	if !ok {
		t.Fatalf("expected APIError, got %T: %v", err, err)
	}
	if apiErr.Code != "DOMAIN_NOT_FOUND" {
		t.Fatalf("unexpected code: %s", apiErr.Code)
	}
	if apiErr.Status != http.StatusNotFound {
		t.Fatalf("unexpected status: %d", apiErr.Status)
	}
}

File: sdk/node/src/client.test.ts
import { describe, it, expect, beforeAll, afterAll, afterEach } from 'vitest'
import { setupServer } from 'msw/node'
import { http, HttpResponse } from 'msw'
import { HnsXClient, APIError, streamSessionEvents } from './client.js'

const server = setupServer()

beforeAll(() => server.listen({ onUnhandledRequest: 'warn' }))
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

function createSSEResponse(chunks: string[]) {
  const encoder = new TextEncoder()
  return new ReadableStream({
    start(controller) {
      for (const chunk of chunks) {
        controller.enqueue(encoder.encode(chunk))
      }
      controller.close()
    },
  })
}

describe('HnsXClient', () => {
  const client = new HnsXClient({ baseURL: 'http://localhost:50052' })

  it('lists domains', async () => {
    server.use(
      http.get('*/api/v1/domains', () => {
        return HttpResponse.json({
          items: [{ id: 'customer-service', version: '1.0.0', status: 'active' }],
          total: 1,
        })
      }),
    )
    const res = await client.domains.list()
    expect(res.total).toBe(1)
    expect(res.items[0].id).toBe('customer-service')
  })

  it('gets a domain', async () => {
    server.use(
      http.get('*/api/v1/domains/customer-service', () => {
        return HttpResponse.json({ id: 'customer-service', version: '1.0.0' })
      }),
    )
    const domain = await client.domains.get('customer-service')
    expect(domain.id).toBe('customer-service')
  })

  it('registers a domain from yaml', async () => {
    server.use(
      http.post('*/api/v1/domains', async ({ request }) => {
        const body = await request.text()
        expect(body).toContain('id: customer-service')
        expect(request.headers.get('content-type')).toContain('yaml')
        return HttpResponse.json({ id: 'customer-service', version: '1.0.0' })
      }),
    )
    const domain = await client.domains.registerYaml('id: customer-service\nversion: "1.0.0"')
    expect(domain.id).toBe('customer-service')
  })

  it('triggers a session', async () => {
    server.use(
      http.post('*/api/v1/sessions', async ({ request }) => {
        const body = (await request.json()) as Record<string, unknown>
        expect(body.domain_id).toBe('customer-service')
        expect((body.trigger as Record<string, unknown>).question).toBe('hi')
        return HttpResponse.json({ id: 'sess-123', state: 'running' })
      }),
    )
    const session = await client.sessions.trigger({
      domainId: 'customer-service',
      trigger: { question: 'hi' },
    })
    expect(session.id).toBe('sess-123')
    expect(session.state).toBe('running')
  })

  it('lists sessions with query params', async () => {
    server.use(
      http.get('*/api/v1/sessions', ({ request }) => {
        const url = new URL(request.url)
        expect(url.searchParams.get('domain')).toBe('customer-service')
        expect(url.searchParams.get('limit')).toBe('10')
        return HttpResponse.json({ items: [{ id: 's1', state: 'completed' }], total: 1 })
      }),
    )
    const res = await client.sessions.list({ domain: 'customer-service', limit: 10 })
    expect(res.items[0].id).toBe('s1')
  })

  it('approves an approval', async () => {
    server.use(
      http.post('*/api/v1/approvals/approve-1/approve', () => {
        return HttpResponse.json({ id: 'approve-1', status: 'approved' })
      }),
    )
    const approval = await client.approvals.approve('approve-1')
    expect(approval.status).toBe('approved')
  })

  it('throws APIError on failure', async () => {
    server.use(
      http.get('*/api/v1/domains/missing', () => {
        return HttpResponse.json({ code: 'DOMAIN_NOT_FOUND', message: 'not found' }, { status: 404 })
      }),
    )
    await expect(client.domains.get('missing')).rejects.toBeInstanceOf(APIError)
    try {
      await client.domains.get('missing')
    } catch (err) {
      expect((err as APIError).code).toBe('DOMAIN_NOT_FOUND')
      expect((err as APIError).status).toBe(404)
    }
  })

  it('streams session events', async () => {
    server.use(
      http.get('*/api/v1/sessions/sess-123/events', () => {
        return new HttpResponse(
          createSSEResponse([
            'event: observation.text\n',
            'data: {"message":"hello"}\n',
            '\n',
            'event: session.completed\n',
            'data: {}\n',
            '\n',
          ]),
          { headers: { 'content-type': 'text/event-stream' } },
        )
      }),
    )

    const events: { name: string; payload: string }[] = []
    for await (const ev of client.streamSessionEvents('sess-123')) {
      events.push(ev)
    }
    expect(events).toHaveLength(2)
    expect(events[0].name).toBe('observation.text')
    expect(events[1].name).toBe('session.completed')
  })
})

describe('streamSessionEvents', () => {
  it('parses standalone helper', async () => {
    server.use(
      http.get('*/api/v1/sessions/standalone/events', () => {
        return new HttpResponse(
          createSSEResponse(['event: ping\n', 'data: {}\n', '\n']),
          { headers: { 'content-type': 'text/event-stream' } },
        )
      }),
    )
    const events: { name: string; payload: string }[] = []
    for await (const ev of streamSessionEvents({
      baseURL: 'http://localhost:50052',
      sessionId: 'standalone',
    })) {
      events.push(ev)
    }
    expect(events).toHaveLength(1)
    expect(events[0].name).toBe('ping')
  })
})
